osquery/events/windows/etw/etw_publisher_processes.cpp

```diff
@@ -0,0 +1,552 @@
+            procStartCacheData->MandatoryLabelSid);
+        procStartData->ProcessSequenceNumber =
+            procStartCacheData->ProcessSequenceNumber;
+        procStartData->ParentProcessSequenceNumber =
+            procStartCacheData->ParentProcessSequenceNumber;
+        procStartData->TokenElevationType =
+            procStartCacheData->TokenElevationType;
+        procStartData->TokenIsElevated = procStartCacheData->TokenIsElevated;
+        shouldDispatch = true;
+      }
+
+      if (shouldDispatch) {
+        // Event is ready to be dispatched
+
+        // Event enrichment phase
+        updateHardVolumeWithLogicalDrive(procStartData->ImageName);
+        updateUserInfo(procStartData->UserSid, procStartData->UserName);
+        updateTokenInfo(procStartData->TokenElevationType,
+                        procStartData->TokenElevationTypeInfo);
+
+        // Event dispatch
+        event_context->data = std::move(eventData);
+        fire(event_context);
+
+        // Remove it from the process start aggregation cache
+        processStartAggregationCache_.erase(processCacheIt);
+
+        // Houskeeping of expired aggregation cache entries
+        cleanOldAggregationCacheEntries();
+
+        // Caching image full path
+        processImageCache_.insert({searchKey, procStartData->ImageName});
+      }
+    }
+  }
+}
+
+void EtwPublisherProcesses::initializeHardVolumeConversions() {
+  const auto& validDriveLetters = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+
+  for (auto& driveLetter : validDriveLetters) {
+    std::string queryPath;
+    queryPath.push_back(driveLetter);
+    queryPath.push_back(':');
+
+    char hardVolume[MAX_PATH + 1] = {0};
+    if (QueryDosDeviceA(queryPath.c_str(), hardVolume, MAX_PATH)) {
+      hardVolumeDrives_.insert({hardVolume, queryPath});
+    }
+  }
+}
+
+void EtwPublisherProcesses::cleanOldAggregationCacheEntries() {
+  // Time stamp value is expressed in 100 nanosecond units, this is about
+  // 10000000 nanoseconds per second
+  static constexpr LONGLONG expiredTime10secs = 10000000 * 10;
+
+  if (processStartAggregationCache_.empty()) {
+    return;
+  }
+
+  FILETIME currentFiletimeTimestamp{0};
+  GetSystemTimeAsFileTime(&currentFiletimeTimestamp);
+
+  ULARGE_INTEGER currentTimestamp{0};
+  currentTimestamp.HighPart = currentFiletimeTimestamp.dwHighDateTime;
+  currentTimestamp.LowPart = currentFiletimeTimestamp.dwLowDateTime;
+
+  auto it = processStartAggregationCache_.begin();
+  while (it != processStartAggregationCache_.end()) {
+    ULARGE_INTEGER eventTimestamp{0};
+    eventTimestamp.HighPart = it->second->CreateTime.dwHighDateTime;
+    eventTimestamp.LowPart = it->second->CreateTime.dwLowDateTime;
+
+    // check if event is still there after 10 secs
+    if ((eventTimestamp.QuadPart + expiredTime10secs) <
+        currentTimestamp.QuadPart) {
+      // event expire and should be deleted
+      processStartAggregationCache_.erase(it++);
+    } else {
+      ++it;
+    }
```

Comment:
Is this the same? Feels easier to understand for me to have the iterator advance outside the conditional
```suggestion
      processStartAggregationCache_.erase(it);
    }
    ++it;
```

Reply:
Thanks for the suggestion, I think the code is more readable now. I'm going to try out this change and see how it goes